### PR TITLE
Global config generator for gem

### PR DESCRIPTION
**Pull request by dmitrijivanchenko**
## Description

Add a new generator with template for global uploadcare widget config

## Checklist

- [ ] Tests (if applicable)
- [ ] Documentation (if applicable)
- [ ] Changelog stub (or use [conventional commit messages](https://www.conventionalcommits.org/))

### CODE REVIEW

File: lib/generators/templates/uploadcare_config_template.erb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  # The signature is an HMAC/SHA256 hex-encoded hash that includes secret_key and expire strings.
+  # To use Secure Uploading for your project, you need to make a signature on your back end using your
+  # project's Secret Key and provide it to File Uploader.
+  # Learn how to make your signature in docs - https://uploadcare.com/docs/security/secure-uploads/
+  # config.secure_signature = ''
+
+  # Stands for the Unix time to which the signature is valid, e.g., 1454902434.
+  # Works together with the secure signature option that defines a period of your signature validity.
+  # config.secure_expire = 1454902434
```
Comment (dmitry-mukhin) [inline lib/generators/templates/uploadcare_config_template.erb:84]: this definitely does not look like a proper conf setting
it should be lifetime, not expire.

File: lib/generators/templates/uploadcare_config_template.erb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  # Defines the file uploader manual crop behavior (see https://uploadcare.com/docs/uploads/file-uploader/#crop-option).
+  # When uploading images, your users can select a crop area.
+  # This option does not force your file uploader to accept images only.
+  # The option also works in the multi-file mode since version 2.3.0.
+  # config.crop = 'disabled'
+
+  # Saves traffic and storage space by resizing images on a client before uploading.
+  # See the detailed option description for details. Using the client-side resize does not force the images only option.
+  # config.image_shrink = ''
```
Comment (dmitry-mukhin) [inline lib/generators/templates/uploadcare_config_template.erb:20]: should be false or string. empty string is bad semantic

File: lib/generators/templates/uploadcare_config_template.erb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  # config.locale_translations = {}
+
+  # Defines pluralization options. See https://uploadcare.com/docs/uploads/file-uploader/#pluralization
+  # config.locale_pluralize = {}
+
+  # The signature is an HMAC/SHA256 hex-encoded hash that includes secret_key and expire strings.
+  # To use Secure Uploading for your project, you need to make a signature on your back end using your
+  # project's Secret Key and provide it to File Uploader.
+  # Learn how to make your signature in docs - https://uploadcare.com/docs/security/secure-uploads/
+  # config.secure_signature = ''
```
Comment (dmitry-mukhin) [inline lib/generators/templates/uploadcare_config_template.erb:80]: this definitely does not look like a proper conf setting
it should be calculated in runtime

**Review by dmitry-mukhin — CHANGES_REQUESTED** [review]

File: lib/generators/templates/uploadcare_config_template.erb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  # config.image_shrink = ''
+
+  # Allows user to remove uploaded files from the file uploader.
+  # Please note that those files are not deleted from your account.
+  # config.clearable = true
+
+  # Allows you to define which upload sources you want to use with the file uploader.
+  # The value is represented by a space-separated ordered list of upload source names.
+  # See https://uploadcare.com/docs/uploads/file-uploader/#upload-sources for details.
+  # config.tabs = 'url file facebook'
```
Comment (evgeniy-r) [inline lib/generators/templates/uploadcare_config_template.erb:29]: Why not an array?

File: lib/generators/templates/uploadcare_config_template.erb [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  # because your users will still be able to choose any file via drag&drop or from URL.
+  # config.input_accept_types = 'image/*'
+
+  # Defines the list of preferred MIME types.
+  # The list should be ordered and contain space-separated MIME types.
+  # Common parts can be marked with asterisks: image/* application/vnd.openxmlformats-officedocument.*.
+  # If no MIME types match the criteria or preferred types aren’t set, default formats are used.
+  # Keep in mind that some cloud services can export data in different formats.
+  # For example, Google document can be exported as Word document, PDF or plain text.
+  # config.preferred_types = 'image/* application/vnd.openxmlformats-officedocument.*'
```
Comment (evgeniy-r) [inline lib/generators/templates/uploadcare_config_template.erb:47]: Why not an array?

File: lib/generators/templates/uploadcare_config_template.erb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  # Learn how to make your signature in docs - https://uploadcare.com/docs/security/secure-uploads/
+  # config.secure_signature = ''
+
+  # Stands for the Unix time to which the signature is valid, e.g., 1454902434.
+  # Works together with the secure signature option that defines a period of your signature validity.
+  # config.secure_expire = 1454902434
+
+  # The option can be used with Authenticated URLs. Defines your proxy backend URL.
+  # See https://uploadcare.com/docs/security/secure-delivery/#proxy-backend.
+  # config.preview_proxy = ''
```
Comment (evgeniy-r) [inline lib/generators/templates/uploadcare_config_template.erb:88]: The empty string again.

File: lib/generators/templates/uploadcare_config_template.erb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+  # The option can be used with Authenticated URLs.
+  # Defines the function that specifies which URL a file uploader should use for image previews.
+  # Function signature is (originalUrl, fileInfo) => previewUrl.
+  # If this option is specified, option previewProxy will be ignored.
+  # config.preview_url_callback = {}
```
Comment (evgeniy-r) [inline lib/generators/templates/uploadcare_config_template.erb:94]: If this is a function, what the hash means?

File: lib/generators/templates/uploadcare_config_template.erb [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+  # Allows you to adjust the quality of an audio recorded via the file uploader Camera Tab.
+  # Refer the link to learn more - https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/MediaRecorder.
+  # config.audio_bits_per_second = 1
+
+  # Allows you to specify the MIME types.
+  # The first supported by the browser will be used as an option of
+  # video stream captured via the file uploader Camera Tab.
+  # Refer the link to learn more - https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/MediaRecorder.
+  # config.video_preferred_mime_types = ''
```
Comment (evgeniy-r) [inline lib/generators/templates/uploadcare_config_template.erb:121]: Array?

File: uploadcare-rails.gemspec [hunk 9]
```diff
@@ -1,29 +1,42 @@
...
-  spec.required_ruby_version = Gem::Requirement.new(">= 2.3.0")
+# Maintain your gem's version:
+require 'uploadcare/rails/version'
 
-  spec.metadata["allowed_push_host"] = "TODO: Set to 'http://mygemserver.com'"
+# Describe your gem and declare its dependencies:
+Gem::Specification.new do |gem|
+  gem.name          = 'uploadcare-rails'
+  gem.authors       = ["@dmitrijivanchenko (Dmitrij Ivanhcneko)"]
+  gem.email         = ['hello@uploadcare.com']
```
Comment (evgeniy-r) [inline uploadcare-rails.gemspec:11]: Is this OK?

File: uploadcare-rails.gemspec [hunk 10]
```diff
@@ -1,29 +1,42 @@
...
-  # The `git ls-files -z` loads the files in the RubyGem that have been added into git.
-  spec.files         = Dir.chdir(File.expand_path('..', __FILE__)) do
-    `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
-  end
-  spec.bindir        = "exe"
-  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
-  spec.require_paths = ["lib"]
+  gem.add_development_dependency 'bundler', '~> 1.6'
+  gem.add_development_dependency 'rake'
+  gem.add_development_dependency 'sqlite3'
```
Comment (evgeniy-r) [inline uploadcare-rails.gemspec:32]: Why do we need this?

File: lib/uploadcare-rails.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+require 'uploadcare/configuration'
+
+module Uploadcare
+  module Rails
```
Comment (evgeniy-r) [inline lib/uploadcare-rails.rb:4]: Can we use this configuration style for other frameworks?
Is this Rails-specific?

File: lib/uploadcare/configuration.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+require 'singleton'
+
+module Uploadcare
+  class Configuration
```
Comment (evgeniy-r) [inline lib/uploadcare/configuration.rb:4]: Should we move this class to the gem?

File: lib/uploadcare-rails.rb (hunk 11 above)
Comment (dmitrijivanchenko) [reply lib/uploadcare-rails.rb:4]: I have not worked with other frameworks, so hard to say if we can use this configuration style for them. At least I guess it framework specific and this is a reason not to place it to the ruby-gem

File: lib/uploadcare/configuration.rb (hunk 12 above)
Comment (dmitrijivanchenko) [reply lib/uploadcare/configuration.rb:4]: Same here. I think that this is framework specific and this is a reason not to place it to the ruby-gem.

File: uploadcare-rails.gemspec (hunk 9 above)
Comment (dmitry-mukhin) [reply uploadcare-rails.gemspec]: ok

File: uploadcare-rails.gemspec (hunk 10 above)
Comment (dmitry-mukhin) [reply uploadcare-rails.gemspec]: most likely for running tests against example/test projects.

**Review by dmitry-mukhin — CHANGES_REQUESTED** [review]

File: lib/generators/templates/uploadcare_config_template.erb [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+  # Sets custom localization options. See https://uploadcare.com/docs/uploads/file-uploader/#localization
+  # config.locale_translations = {}
+
+  # Defines pluralization options. See https://uploadcare.com/docs/uploads/file-uploader/#pluralization
+  # config.locale_pluralize = {}
+
+  # Stands for the Unix time to which the signature is valid, e.g., 1454902434.
+  # Works together with the secure signature option that defines a period of your signature validity.
+  # config.signature_lifetime = 1454902434
```
Comment (dmitry-mukhin) [inline lib/generators/templates/uploadcare_config_template.erb:78]: why is this still here?

File: uploadcare-rails.gemspec [hunk 14]
```diff
@@ -1,29 +1,39 @@
...
-  spec.required_ruby_version = Gem::Requirement.new(">= 2.3.0")
+# Maintain your gem's version:
+require 'uploadcare/rails/version'
 
-  spec.metadata["allowed_push_host"] = "TODO: Set to 'http://mygemserver.com'"
+# Describe your gem and declare its dependencies:
+Gem::Specification.new do |gem|
+  gem.name          = 'uploadcare-rails'
+  gem.authors       = ["@dmitrijivanchenko (Dmitrij Ivanhcneko)"]
+  gem.email         = ['dmitrijivanchenko@gmail.com']
```
Comment (dmitry-mukhin) [inline uploadcare-rails.gemspec:11]: please leave company's address here:
https://guides.rubygems.org/specification-reference/#email

**Review by dmitry-mukhin — CHANGES_REQUESTED** [review]

**Pull request merged**